src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs

```diff
@@ -205,9 +206,55 @@ private ImmutableArray<ISymbol> GetSymbolsForTypeOrNamespaceContext()
                 return symbols.WhereAsArray(s => !s.IsDelegateType() && !s.IsInterfaceType());
             }
 
+            if (_context.IsAsyncMemberDeclarationContext)
+            {
+                return symbols.WhereAsArray(IsSymbolValidForAsyncDeclarationContext);
+            }
+
             return symbols;
         }
 
+        private bool IsSymbolValidForAsyncDeclarationContext(ISymbol symbol)
+        {
+            if (symbol.IsNamespace())
+            {
+                return true;
+            }
+
+            if (symbol is not INamedTypeSymbol namedType ||
+                symbol.IsDelegateType() ||
+                namedType.IsEnumType())
+            {
+                return false;
+            }
+
+            if (namedType.TypeKind == TypeKind.Interface)
+            {
+                // The only interfaces, that are valid in async context are IAsyncEnumerable and IAsyncEnumerator.
+                // So if we are validating an interface, then we can just check for 2 of this possible variants
+                return (namedType.Name == "IAsyncEnumerable" || namedType.Name == "IAsyncEnumerator") &&
+                    namedType.TypeParameters.Length == 1 &&
+                    namedType.IsFromSystemRuntimeOrMscorlibAssembly();
+            }
+
+            if (namedType.TypeKind == TypeKind.Class &&
+                namedType.Name == "Task" &&
```

Comment:
nameof(Task).  Also, i think we have extensions to get the Task types like `compilation.TaskOfTType()`.

Reply:
`nameof` acquired. However, I didn't find any `TaskOfTType` extension method. Searched in the solution with `TaskOf` query